repair/repair.cc

```diff
@@ -2052,6 +2052,77 @@ future<> repair_service::replace_with_repair(locator::token_metadata_ptr tmptr,
     co_return co_await do_rebuild_replace_with_repair(std::move(cloned_tmptr), std::move(op), myloc.dc, reason, std::move(ignore_nodes));
 }
 
+// Repair all tablets belong to this node for the given table
+future<> repair_service::repair_tablets(repair_uniq_id rid, sstring keyspace_name, sstring table_name, host2ip_t host2ip, bool primary_replica_only) {
+    auto& t = _db.local().find_column_family(keyspace_name, table_name);
+    if (!t.uses_tablets()) {
+        throw std::runtime_error(format("Table {}.{} is not a tablet table", keyspace_name, table_name));
+    }
+    table_id tid = t.schema()->id();
+    // FIXME: we need to wait for current tablet movement and disable future tablet movment
+    auto erm = t.get_effective_replication_map();
+    auto& tmap = erm->get_token_metadata_ptr()->tablets().get_tablet_map(tid);
+    struct repair_tablet_meta {
+        locator::tablet_id id;
+        dht::token_range range;
+        locator::host_id master_host_id;
+        shard_id master_shard_id;
+        locator::tablet_replica_set replicas;
+    };
+    std::vector<repair_tablet_meta> metas;
+    auto myhostid = erm->get_token_metadata_ptr()->get_my_id();
+    auto myip = erm->get_topology().my_address();
+    size_t total_tablets = 0;
+    size_t nr = 0;
+    co_await tmap.for_each_tablet([&] (locator::tablet_id id, const locator::tablet_info& info) {
+        total_tablets++;
+        auto range = tmap.get_token_range(id);
+        auto& replicas = info.replicas;
+        bool found = false;
+        shard_id master_shard_id;
+        // Repair all tablets belong to this node
+        for (auto& r : replicas) {
+            if (r.host == myhostid) {
+                master_shard_id = r.shard;
+                found = true;
+                break;
+            }
+            if (primary_replica_only) {
+                break;
```

Comment:
So I think this is wrong for multi-datacenter deployments.

Reply:
Maybe: `for (auto& r: replicas | filtered(on_this_datacenter)) {`

Reply:
Yes, see https://github.com/scylladb/scylladb/pull/16600#discussion_r1440008392. By default repair is global repair.

Reply:
Ah, I misunderstood -pr (but could have derived it using logic).